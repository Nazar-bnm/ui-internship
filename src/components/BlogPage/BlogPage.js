import React from 'react';
import HttpService from '../../service/HttpService/httpService';

import BlogItem from './BlogItem';
import { Button } from '@/shared';
import Labels from '../Labels';
import { formatDate } from '../../helpers';

import './BlogPage.scss';

const CN = 'blog-page';

class BlogPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currentRangeStart: 0,
      currentRangeEnd: 3,
      isButtonDisabled: false
    };

    this.getOlderPosts = this.getOlderPosts.bind(this);
  }

  componentDidMount() {
    this.getBlogItems();
  }

  async getBlogItems() {
    const userAPI = new HttpService();
    const blogsURL = `${process.env.SERVER_URL}/blogs`;
    const { getBlogItemsSuccess, getBlogItemsError } = this.props;

    try {
      const response = await userAPI.get(blogsURL);

      if (response.status === 404) {
        throw Error(response.statusText);
      }
      getBlogItemsSuccess(response.data);
    } catch (error) {
      getBlogItemsError(error);
    }
  }

  getItemsToRender() {
    const { currentRangeStart, currentRangeEnd } = this.state;
    const { blogItems } = this.props;
    const blogItemsToRender = blogItems.slice(currentRangeStart, currentRangeEnd);
    const maxSymbolsNumber = 301;

    return blogItemsToRender.map((el, index) => {
      const shortText = `${el.description.slice(0, maxSymbolsNumber)}...`;
      const imageSrc = `${process.env.BLOG_IMAGE_URL}/${el.photo}`;
      const {
        _id,
        title,
        date,
        labels
      } = el;

      const formattedDate = formatDate(date);

      return (
        <BlogItem
          key={_id}
          title={title}
          date={formattedDate}
          photo={imageSrc}
          text={shortText}
          labels={labels}
          className={index === 2 ? null : 'bottom-border'}
        />
      );
    });
  }

  getOlderPosts() {
    const blogItemsPerStep = 3;
    const { currentRangeStart, currentRangeEnd } = this.state;
    const { blogItems } = this.props;
    const itemsNumber = blogItems.length;
    const nextRangeEnd = currentRangeEnd + blogItemsPerStep;

    this.setState({
      currentRangeStart: currentRangeStart + blogItemsPerStep,
      currentRangeEnd: currentRangeEnd + blogItemsPerStep
    });

    window.scrollTo(0, 0);

    if (nextRangeEnd > itemsNumber) {
      this.setState({
        isButtonDisabled: true
      });
    }
  }

  render() {
    const { isButtonDisabled } = this.state;
    const { blogItems, getLabel } = this.props;

    return (
      <div className={`${CN} content`}>
        <div className="categories">
          <p>Categories</p>
          <hr />
          <p>Celebrity style (39)</p>
          <p>Fashion shows (15)</p>
          <p>Shopping (27)</p>
          <p>Shopping (119)</p>
        </div>
        <div className="labels__wrapper">
          <Labels blogItems={blogItems} getLabel={getLabel} />
        </div>
        <div className={`${CN}__items-wrapper`}>
          {this.getItemsToRender()}
          <Button
            disabled={isButtonDisabled}
            className="transparent-button"
            onClick={this.getOlderPosts}
          >
            Older posts &gt;
          </Button>
        </div>
      </div>
    );
  }
}

export default BlogPage;
