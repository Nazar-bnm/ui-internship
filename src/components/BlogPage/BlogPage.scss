@import '../../scss/global_styles';

.blog-page {
  @include flex($direction: column);
  padding: 10px 0;

  .categories {
    border: 1px solid $border-lightGrey;
    padding: 10px;
  }

  .transparent-button {
    align-self: flex-end;
  }

  &__items-wrapper {
    @include flex($direction: column);
  }

  .labels__wrapper {
    display: none;
  }

  &__image-wrapper {
    position: relative;
  }
}

@media only screen and (min-width: 1024px) {
  .blog-page {
    display: grid;
    grid-template-columns: 2fr 1fr;
    grid-column-gap: 20px;
    grid-row-gap: 20px;

    .categories {
      grid-column: 2;
      grid-row: 1;
      margin: 20px 0;
    }

    .labels__wrapper {
      display: block;
      grid-column: 2;
      grid-row: 2;
      padding: 25px 15px;
      border: 1px solid #ddd;
    }

    &__items-wrapper {
      @include flex($direction: column, $wrap: nowrap);
      grid-column: 1;
      grid-row: 1/4;
    }
  }
}