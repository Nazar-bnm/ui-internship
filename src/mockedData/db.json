{
  "employees": [
    {
      "id": 1,
      "first_name": "Sebastian",
      "last_name": "Eschweiler",
      "email": "sebastian@codingthesmartway.com"
    },
    {
      "id": 2,
      "first_name": "Steve",
      "last_name": "Palmer",
      "email": "steve@codingthesmartway.com"
    },
    {
      "id": 3,
      "first_name": "Ann",
      "last_name": "Smith",
      "email": "ann@codingthesmartway.com"
    }
  ],
  "what-is-new" : [
    {
      "id": 1,
      "photo": "https://www.next.ua/nxtcms/resource/image/1889302/portrait_ratio1x1/525/525/2fccc457347d46d0c506ee9ea0396ae8/of/emma-willis.jpg",
      "title": "Black and White",
      "collection": "spring collection 2016",
      "category": "women"
    },
    {
      "id": 2,
      "photo": "https://www.next.ua/nxtcms/resource/image/1233600/portrait_ratio1x1/525/525/69ca86774e073288bc792479650a14cc/Jn/all-baby-girl.jpg",
      "title": "Color Summer",
      "collection": "spring collection 2016",
      "category": "kids"
    },
    {
      "id": 3,
      "photo": "https://www.next.ua/nxtcms/resource/image/1915752/portrait_ratio1x1/525/525/1cf45576cf1de5e2bab187220d4ebeb0/ds/g83-mb-sh819338-lm-357.jpg",
      "title": "Vintage For Him",
      "collection": "spring collection 2016",
      "category": "men"
    }
  ],
  "most-popular": [
    {
      "id": 1,
      "image": "https://i.dailymail.co.uk/1s/2018/12/15/12/7458890-6498875-image-a-2_1544876240893.jpg",
      "label": "new",
      "title": "Black dress",
      "price": "99,50$",
      "sizes": "S - M - L"
    },
    {
      "id": 2,
      "image": "https://i.pinimg.com/564x/99/5d/f3/995df3032c828381332af1dea4e5c590.jpg",
      "label": "sale",
      "title": "Red evening dress",
      "price": "179.00$",
      "sizes": "S - M - L"
    },
    {
      "id": 3,
      "image": "https://i.redd.it/grla2qtpa5721.jpg",
      "label": "",
      "title": "Fit & flare dress",
      "price": "99.00$",
      "sizes": "S - M - L"
    }
  ]
}
