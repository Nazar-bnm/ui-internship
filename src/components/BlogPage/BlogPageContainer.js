import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { getBlogItemsSuccess, getBlogItemsError, getLabel } from '../../actions/actionsBlogPage';

import BlogPage from './BlogPage';

const mapStateToProps = (state) => ({
  blogItems: state.reducerBlogPage.blogItems,
  error: state.reducerBlogPage.error,
  label: state.reducerBlogPage.label
});

const mapDispatchToProps = (dispatch) => ({
  getBlogItemsSuccess: bindActionCreators(getBlogItemsSuccess, dispatch),
  getBlogItemsError: bindActionCreators(getBlogItemsError, dispatch),
  getLabel: bindActionCreators(getLabel, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);
