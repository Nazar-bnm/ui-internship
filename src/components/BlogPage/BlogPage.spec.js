import React from 'react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { mount } from 'enzyme';

import BlogPage from './BlogPage';
import mockedData from '../../mockedDataForTests';

describe('<BlogPage />', () => {
  const { mockedBlogsPage } = mockedData;
  let wrapper;

  beforeEach(() => {
    const mock = new MockAdapter(axios);

    wrapper = mount(<BlogPage blogItems={mockedBlogsPage} />);
    mock.onGet(`${process.env.BASE_URL}/blogsPage`).reply(200, mockedBlogsPage);
  });

  test('should check `componentDidMount()`', () => {
    const instance = wrapper.instance();
    
    jest.spyOn(instance, 'getBlogItems');
    instance.componentDidMount();
    expect(instance.getBlogItems).toHaveBeenCalledTimes(1);
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
