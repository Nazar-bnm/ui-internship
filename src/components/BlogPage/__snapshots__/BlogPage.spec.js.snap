// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlogPage /> should match snapshot 1`] = `
<div
  className="blog-page content"
>
  <div
    className="categories"
  >
    <p>
      Categories
    </p>
    <p>
      Celebrity style (39)
    </p>
    <p>
      Fashion shows (15)
    </p>
    <p>
      Shopping (27)
    </p>
    <p>
      Shopping (119)
    </p>
  </div>
  <div
    className="labels__wrapper"
  >
    <Labels
      blogItems={
        Array [
          Object {
            "date": "12 october 2019",
            "description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus. Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna. Sed consequat, leo eget bibendum sodales, augue velit cursus nunc,",
            "id": 1,
            "intro": "we know what you want this summer. most fashion looks",
            "labels": Array [
              "summer",
              "swimsuit",
              "bags",
              "sunglasses",
            ],
            "photo": "1.jpg",
            "title": "summer trends",
          },
          Object {
            "date": "11 january 2018",
            "description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus. Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna. Sed consequat, leo eget bibendum sodales, augue velit cursus nunc,",
            "id": 2,
            "intro": "new trends comming through",
            "labels": Array [
              "boho",
              "summer",
              "trend",
              "style",
            ],
            "photo": "2.jpg",
            "title": "Free-spirited style for boho-chic",
          },
          Object {
            "date": "25 may 2019",
            "description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus. Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna. Sed consequat, leo eget bibendum sodales, augue velit cursus nunc,",
            "id": 3,
            "intro": "military style boots",
            "labels": Array [
              "boots",
              "autumn",
              "trend",
              "style",
              "military",
            ],
            "photo": "3.jpg",
            "title": "military style",
          },
        ]
      }
      getLabel={[MockFunction]}
    />
  </div>
  <div
    className="blog-page__items-wrapper"
  >
    <BlogItem
      className="bottom-border"
      date="12 october 2019"
      key="1"
      labels={
        Array [
          "summer",
          "swimsuit",
          "bags",
          "sunglasses",
        ]
      }
      photo="undefined/1.jpg"
      text="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec p..."
      title="summer trends"
    />
    <BlogItem
      className="bottom-border"
      date="11 january 2018"
      key="2"
      labels={
        Array [
          "boho",
          "summer",
          "trend",
          "style",
        ]
      }
      photo="undefined/2.jpg"
      text="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec p..."
      title="Free-spirited style for boho-chic"
    />
    <BlogItem
      className={null}
      date="25 may 2019"
      key="3"
      labels={
        Array [
          "boots",
          "autumn",
          "trend",
          "style",
          "military",
        ]
      }
      photo="undefined/3.jpg"
      text="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec p..."
      title="military style"
    />
    <Button
      className="transparent-button"
      disabled={false}
      icon={null}
      onClick={[Function]}
    >
      Older posts &gt;
    </Button>
  </div>
</div>
`;
